Multilevel mailbox structure

dehnhardt opened the pull request:
- [ ] (Soft) requires #3466
- [x] Logic for `addAccount`
- [x] Logic for `addMailbox`
- [x] ~~Edge case: removal of a mailbox that has children -> children should be promoted by one level (if IMAP even allow this operation)~~ -> problematic on IMAP level as well, so we're hiding the *delete* action for mailboxes with children.
- [x] Detect and persist personal namespace (related to https://github.com/nextcloud/mail/issues/3272, but fully automated)

Fixes #794 

Signed-off-by: Holger Dehnhardt <holger@dehnhardt.org>
With this pr it is possible to display a folder structure of any depth.
The structure is only created using the getter in the frontend. 
Creating new folders within the structure works - but is not tested carefully.
There are some problems with the display. 
* Replacing the folder icons with the arrows does not work properly 
![image](https://user-images.githubusercontent.com/5556832/82752828-154f7800-9dc1-11ea-81f0-45ccc11ffc70.png)

* the navigation area quickly becomes too narrow.
See comments inside the changes...
![image](https://user-images.githubusercontent.com/5556832/82688684-925fdd80-9c59-11ea-8baf-6abdbcdc41a8.png)

_To make the goal of this PR clearer, this picture shows how the structure is presented so far:_
![image](https://user-images.githubusercontent.com/5556832/85917870-2aa34080-b85e-11ea-9050-4e88b95be6f3.png)

dehnhardt commented on lib/Folder.php:
```diff
@@ -62,6 +68,9 @@ public function __construct(int $accountId, Horde_Imap_Client_Mailbox $mailbox,
+		$lastSeparator = strrpos($mailbox->utf8, $delimiter) > 0 ? strrpos($mailbox->utf8, $delimiter)+1: 0;
+		$this->displayName = substr($mailbox->utf8, $lastSeparator);
+		$this->path = substr($mailbox->utf8, 0, $lastSeparator);
 	}
```
Seperate display name and path. The path is used to select the subfolders.

ChristophWurst replied:
we only need it on the front-end. can we just generate this there?

dehnhardt replied:
Yes, we can;-), separating path and file name and creating the hierarchy on the fly is ok so far?

ChristophWurst replied:
I'm not entirely sure. But I did not fully grasp the intended change so far

dehnhardt commented on src/store/getters.js:
```diff
@@ -38,12 +38,16 @@ export const getters = {
-
-		return folder.folders.map((id) => state.folders[id])
+	getSubfolders: (state) => (accountId, folderId) => {
+		const folder = state.folders[normalizedFolderId(accountId, folderId)]
```
getters.getFolder(accountId, folderId) does not work here.

dehnhardt commented on src/store/mutations.js:
```diff
@@ -56,7 +56,7 @@ export default {
 
 		// Save the folders to the store, but only keep IDs in the account's folder list
-		const folders = buildMailboxHierarchy(sortMailboxes(account.folders || []), havePrefix(account.folders))
+		const folders = sortMailboxes(account.folders || [])
```
We don't need building a hierarchy here anymore.

jancborchardt commented:
Looks quite nice! To prevent the issue of the navigation getting too narrow, we need to do a less extreme indent for the subentries. That is, the left side of the icon should left-align with the text of the parent folder.

dehnhardt commented:
@jancborchardt I had the same idea. But I wasn't sure if there are design guidelines from Nextcloud that prohibit this kind of changes.
Would the change you proposed have to be made by someone specific to the Nextcloud core, as it could affect many apps?

dehnhardt commented:
I opened an issue in https://github.com/nextcloud/nextcloud-vue for the display problems.
https://github.com/nextcloud/nextcloud-vue/issues/1140

dehnhardt commented:
@ChristophWurst, @GretaD: May I draw your attention again to this ticket for three reasons. 
- the bug #3289 would have been fixed already
- the bug #794 should be fixed - as well as the requirement to display the tree structure correctly
- To implement the Sieve filter interface I need a mailbox picker, which I would like to implement with the correct tree structure.
However, I would like to make sure that the structure as I implemented it is ok.
I know that the last weeks were quite busy because of the release, but maybe you can find time now for a short evaluation.

ChristophWurst commented on src/components/NavigationFolder.vue:
```diff
@@ -33,7 +33,7 @@
 		<!-- actions -->
 		<template slot="actions">
-			<template v-if="top">
+			<template>
```
fixed by #3290

dehnhardt replied:
I know. But I was first;-)

ChristophWurst commented on src/components/NavigationFolder.vue:
```diff
@@ -175,6 +175,7 @@ export default {
 		},
 		subFolders() {
 			return this.$store.getters.getSubfolders(this.account.id, this.folder.id)
+			//return this.folder.folders
```
remove?

dehnhardt replied:
Both done...

ChristophWurst commented:
I'm currently focusing on #2125 but I'll look into this soon.

dehnhardt commented:
> I'm currently focusing on #2125 but I'll look into this soon.

I've seen that. Take your time ;-)

dehnhardt commented:
@ChristophWurst: Now that https://github.com/nextcloud/nextcloud-vue/issues/1140 has come to the mails masterbranch, wouldn't it be wonderful to have this in master as well ;-)
Rebased and squashed.

ChristophWurst commented:
Sure, let's do this :) This requires a few adaptations for #3439 but I can help with that if you like.

ChristophWurst commented on lib/Folder.php:
```diff
@@ -45,6 +45,13 @@ class Folder {
 	private $specialUse;
 
+	/** @var string */
+	private $path;
```
there is no code that write to this, so it will always remain `null`

ChristophWurst commented:
I took the liberty to just rebase the branch, resolve the conflicts and fix some unused code (leftover from previous rebase possibly).

The patch looks clean, but I'm unsure if your attended changes are still here. Could you please check and give me a quick update? We can always reset to an older state if I accidentally dropped some important changes :pray:

ChristophWurst commented:
![Bildschirmfoto von 2020-08-27 14-25-04](https://user-images.githubusercontent.com/1374172/91441922-55ffd700-e871-11ea-96ef-da6d03ffbbb6.png)
![7d5a79a7afb55fe9f7b593a9465cc3f5](https://user-images.githubusercontent.com/1374172/91441925-56986d80-e871-11ea-8986-e235f949b7a1.jpeg)


woooooooorks again :)

dehnhardt commented:
To be honest, I can no longer tell which changes were originally mine and which are yours.
But I checked out and tested the branch on my test system and it seems to work how it should.
I just wonder why the indentation is still so big. Is the current Nextcloud vue version not merged yet?

ChristophWurst commented:
> To be honest, I can no longer tell which changes were originally mine and which are yours.
> But I checked out and tested the branch on my test system and it seems to work how it should.

No worries. I spent some time on this yesterday and also added some unit tests.

> I just wonder why the indentation is still so big. Is the current Nextcloud vue version not merged yet?

It's not, hence the TODO in the PR description.

GretaD reviewed (approved):
tested and works
